net/ucspi-tcp6-0.99a/man/rblsmtpd.1
.TH rblsmtpd 1
.SH NAME
rblsmtpd \- block emails from RBL-listed sites and unwanted sources
.SH SYNOPSIS
.B rblsmtpd
.I opts
.I prog

.SH DESCRIPTION
.B rblsmtpd 
works with any SMTP server that can run under
.BR tcpserver (1).
It accepts a series of getopt-style options
.I opts
while
.I prog
consists of one or more arguments.

Normally
.B rblsmtpd
runs
.IR prog .
.I prog
is expected to carry out an SMTP conversation to receive incoming emails.

However,
.B rblsmtpd
does not invoke
.I prog
if it is told to block emails from this client. 
Instead it carries out its own limited SMTP conversation, 
temporarily rejecting all attempts to send an email. 
Meanwhile it prints one line on descriptor 2 to log its activity.

.B rblsmtpd
drops the limited SMTP conversation after 60 seconds, 
even if the client has not quit by then.

.SH RBL LOOKUP OPTIONS
.TP
.B \-t \fIn
Change the timeout to
.I n
seconds.
.P
.B Blocked clients
.P
If the 
.B RBLSMTPD 
environment variable is set and is nonempty,
.B rblsmtpd
blocks emails. It uses 
.B RBLSMTPD 
as an error message for the client. Normally
.B rblsmtpd
runs under
.BR tcpserver (1);
you can use
.BR tcprules (1)
to set 
.B RBLSMTPD 
for selected clients. 

If 
.B RBLSMTPD
is set and is empty,
.B rblsmtpd
does not block emails. 

If 
.B RBLSMTPD 
is not set,
.B rblsmtpd
looks up 
.I TCPREMOTEIP
in the RBL, and blocks emails if 
.I TCPREMOTEIP
is listed.
.B tcpserver
sets up
.I TCPREMOTEIP 
as the IP address of the remote host.
.TP
.B \-r \fIbase
Use
.I base
as an RBL source. An IPv4 address
.I a.b.c.d
is listed by that source if
.I d.c.b.a.base
has a TXT record.
An IPv6 address is expanded in it's inverse nibble format.
.B rblsmtpd
uses the contents of the TXT record as an error message for the client.
.TP
.B \-a \fIbase
Use
.I base
as an anti-RBL source. An IPv4 address
.I a.b.c.d
is anti-listed by that source if
.I d.c.b.a.base
has an A record. In this case
.B rblsmtpd
does not block emails. The same holds for IPv6 addresses (see below) 
and a corresponding AAAA record. 
.P
You may supply any number of
.B \-r
and
.B \-a
options.
.B rblsmtpd
tries each source in turn until it finds one that lists or anti-lists 
.IR TCPREMOTEIP . 

If you want to run your own RBL source or anti-RBL source for
.BR rblsmtpd ,
you can use
.B rbldns
from the DNScache (djbdns) package.

.SH SMTP REPLIES 
Normally, if 
.B RBLSMTPD 
is set,
.B rblsmtpd
uses a 451 error code in it's limited SMTP conversation. 
This tells legitimate clients to try again later. 
It gives innocent relay operators a chance to see the problem, 
prohibit relaying, get off the RBL, and get the email delivered. 

However, if 
.B RBLSMTPD 
begins with a hyphen ('-'),
.B rblsmtpd
removes the hyphen and uses a 553 error code together with the message.
This tells legitimate clients to bounce the email immediately. 

There are several error-handling options for RBL lookups:
.TP
.B \-B
(Default.) Use a 451 error code for IP addresses listed in the RBL.
.TP
.B \-b
Use a 553 error code for IP addresses listed in the RBL.
.TP
.B \-C
(Default.) Handle RBL lookups in a ``fail-open'' mode. 
If an RBL lookup fails temporarily, assume that the address is not listed; 
if an anti-RBL lookup fails temporarily, assume that the address is anti-listed. 
Unfortunately, a knowledgeable attacker can force an RBL lookup or an anti-RBL 
lookup to fail temporarily, so that his email is not blocked.
.TP
.B \-c
Handle RBL lookups in a ``fail-closed'' mode. 
If an RBL lookup fails temporarily, assume that the address is listed 
(but use a 451 error code even with -b). If an anti-RBL lookup fails temporarily, 
assume that the address is not anti-listed (but use a 451 error code even if a 
subsequent RBL lookup succeeds with -b). Unfortunately, this sometimes delays legitimate emails.


.SH INTERROGATION MODE
.B rblsmtpd
may be used to only query RBLs and to present the results to
.BR qmail-smtpd
in an interrogation mode.
.TP
.B \-i
interrogation mode; the RBL information is written on descriptor 2 and available
via the environment variable
.B RBLSMTPD
for further decisions.

.SH GREETDELEY OPTIONS
Introduce a certain delay in the SMTP connection;
either before or after the RBL, the anti-RBL lookups respectively,
depending whether this option is defined before or after the lookups.
.TP
.B \-w n
provides a delay of 'n' seconds for all connections.
.TP
.B \-W
evaluate the environment variable
.B GREETDELAY
and potentially delay the call to
.BR qmail-smtpd .

.SH IPv4/IPv6 INVERSE NIBBLE FORMAT
.B
rblsmtpd 
constructs from the incoming IP address \fITCPREMOTEIP\fR
by means of it's inverse nibble format the resulting
inverse domain name. If \fIPROTO\fR=\fBTCP\fR
the standard inverse IPv4 address scheme is used, whereas
in case of \fIPROTO\fR=\fBTCP6\fR the IPv6 inverse domain name
is computed based on the expanded IPv6 address:
.IP
fe80::1 =>  
1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.e.f.(tld)
.P
.B rblsmtpd 
recognizes IPv4-mapped IPv6 addresses 
and strips the IPv6 prefix 
piror to build the IPv4 inverse address
for the RBL lookup.

.SH SEE ALSO
tcpserver(1),
tcprules(1),
tcprulescheck(1),
fixcrio(1),
recordio(1),
rblsmtpd(1),
tcpclient(1),
who@(1),
date@(1),
finger@(1),
http@(1),
tcpcat(1),
mconnect(1),
tcp-environ(5)


net/ucspi-tcp6-0.99a/src/rblsmtpd.c
#include "byte.h"
#include "str.h"
#include "scan.h"
#include "fmt.h"
#include "env.h"
#include "exit.h"
#include "sig.h"
#include "buffer.h"
#include "readwrite.h"
#include "sgetopt.h"
#include "strerr.h"
#include "stralloc.h"
#include "commands.h"
#include "pathexec.h"
#include "dns.h"
#include "ip6.h"

#define FATAL "rblsmtpd: fatal: "

void nomem(void)
{
  strerr_die2x(111,FATAL,"out of memory");
}
void usage(void)
{
  strerr_die1x(100,"rblsmtpd: usage: rblsmtpd [ -B ] [ -b ] [ -C ] [ -c ] [ -i ] [ -t timeout ] [ -r base ] [ -a base ] [-W] [-w delay] smtpd [ arg ... ]");
}

char *tcp_proto;
char *ip_env;
static stralloc ip_reverse;
int flagip6 = 0;

static inline char tohex(char c) { return c>=10?c-10+'a':c+'0'; }

void ip_init(void)
{
  unsigned int i;
  unsigned int j;
  char hexval;
  unsigned char remoteip[16];

  tcp_proto = env_get("PROTO");
  if (!tcp_proto) tcp_proto = "";
  ip_env = env_get("TCPREMOTEIP");
  if (!ip_env) ip_env = "";
  if (str_diff(tcp_proto,"TCP6") == 0) {
    if (byte_equal(ip_env,7,V4MAPPREFIX)) 
      ip_env = ip_env + 7;
     else
      flagip6 = 1;
  }
 
  if (!stralloc_copys(&ip_reverse,"")) nomem();

  if (flagip6) { 
    if ((ip6_scan(ip_env,remoteip)) == 0) return;

    for (j = 16; j > 0; j--) {
      hexval = tohex(remoteip[j - 1] & 15);
      if (!stralloc_catb(&ip_reverse,&hexval,1)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem();

      hexval = tohex(remoteip[j - 1] >> 4);
      if (!stralloc_catb(&ip_reverse,&hexval,1)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem(); 
    }
  } else {
    i = str_len(ip_env);
    while (i) {
      for (j = i;j > 0;--j) if (ip_env[j - 1] == '.') break;
      if (!stralloc_catb(&ip_reverse,ip_env + j,i - j)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem();
      if (!j) break;
      i = j - 1;
    }
  }
}

unsigned long timeout = 60;
int flagrblbounce = 0;
int flagfailclosed = 0;
int flagmustnotbounce = 0;
int flagrblinfo = 0;

int decision = 0; /* 0 undecided, 1 accept, 2 direct refuse, 3 direct bounce, 4 rbl refuse, 5 rbl bounce */
static stralloc text; /* defined if decision is > 2 */

static stralloc tmp;

void rbl(char *base)
{
  if (decision) return;
  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
  if (dns_txt(&text,&tmp) == -1) {
    flagmustnotbounce = 1;
    if (flagfailclosed) {
      if (!stralloc_copys(&text,"temporary RBL lookup error")) nomem();
      decision = 4;
    }
    return;
  }
  if (text.len)
    if (flagrblbounce)
      decision = 5;
    else
      decision = 4;
}

void antirbl(char *base)
{
  if (decision) return;
  int flagip;

  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
   
  if (flagip6) 
    flagip = dns_ip6(&text,&tmp);
  else
    flagip = dns_ip4(&text,&tmp);
 
  if (flagip == -1) {
    flagmustnotbounce = 1;
    if (!flagfailclosed)
      decision = 1;
    return;
  }
  if (text.len)
    decision = 1;
}

char strnum[FMT_ULONG];
static stralloc message;
static stralloc info;

char inspace[64]; buffer in = BUFFER_INIT(read,0,inspace,sizeof inspace);
char outspace[1]; buffer out = BUFFER_INIT(write,1,outspace,sizeof outspace);

void wait(unsigned long delay)
{
  unsigned long u;
  char *x;
 
  if (!delay) {
    x = env_get("GREETDELAY");
    if (x) { 
      scan_ulong(x,&u); 
      delay = u; 
    }
  }

  if (delay) {
    if (!stralloc_copys(&info,"greetdelay: ")) nomem();

    buffer_puts(buffer_2,"rblsmtpd: ");
    buffer_puts(buffer_2,ip_env);
    buffer_puts(buffer_2," pid ");
    buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
    buffer_puts(buffer_2,": ");
    buffer_put(buffer_2,info.s,info.len);
    buffer_put(buffer_2,strnum,fmt_ulong(strnum,delay));
    buffer_puts(buffer_2,"\n");
    buffer_flush(buffer_2);

    if (!stralloc_cats(&info,"\r\n")) nomem();

    sleep(delay);
  }
}

void rblinfo(void)
{
  int i; 
 
  if (!stralloc_copys(&message,"[RBL info] ")) nomem();

  if (text.len > 200) text.len = 200;
  if (!stralloc_cat(&message,&text)) nomem();

  for (i = 0;i < message.len;++i)
    if ((message.s[i] < 32) || (message.s[i] > 126))
      message.s[i] = '?';

  buffer_puts(buffer_2,"rblsmtpd: ");
  buffer_puts(buffer_2,ip_env);
  buffer_puts(buffer_2," pid ");
  buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
  buffer_puts(buffer_2,": ");
  buffer_put(buffer_2,message.s,message.len);
  buffer_puts(buffer_2,"\n");
  buffer_flush(buffer_2);

  if (!stralloc_copy(&info,&message)) nomem();
  if (!stralloc_0(&info)) nomem();
  if (!pathexec_env("RBLSMTPD",info.s)) nomem();
}

void reject() { buffer_putflush(&out,message.s,message.len); }
void accept() { buffer_putsflush(&out,"250 rblsmtpd.local\r\n"); }
void greet()  { buffer_putsflush(&out,"220 rblsmtpd.local\r\n"); }
void quit()   { buffer_putsflush(&out,"221 rblsmtpd.local\r\n"); _exit(0); }
void drop()   { _exit(0); }

struct commands smtpcommands[] = {
  { "quit", quit, 0 }
, { "helo", accept, 0 }
, { "ehlo", accept, 0 }
, { "mail", accept, 0 }
, { "rset", accept, 0 }
, { "noop", accept, 0 }
, { 0, reject, 0 }
} ;

void rblsmtpd(void)
{
  int i;
  char *rblmessage;

  rblmessage = env_get("RBLMESSAGE");

  if (flagmustnotbounce || (decision == 2) || (decision == 4)) {
    if (!stralloc_copys(&message,"451 ")) nomem();
  }
  else
    if (!stralloc_copys(&message,"553 ")) nomem();

  if (rblmessage)
    if (!stralloc_cats(&message, rblmessage))
      nomem();

  /* replace %IP% with content of $TCPREMOTEIP in RBLSMTPD string */
  for (i = 0;i <= ( text.len - 4 );++i)
    if (text.s[i] == '%' && text.s[i+1] == 'I' && text.s[i+2] == 'P' && text.s[i+3] == '%') {

      if (!stralloc_copyb(&tmp,text.s,i)) nomem();

      if (!stralloc_cats(&tmp,ip_env)) nomem();

      if (i < ( text.len - 4)) {
	if (!stralloc_catb(&tmp,text.s + i + 4, text.len - i - 4)) nomem();
      }

      if (!stralloc_copy(&text,&tmp)) nomem();

      i = i + 4;
    }


  if (text.len > 200) text.len = 200;
  if (!stralloc_cat(&message,&text)) nomem();
  for (i = 0;i < message.len;++i)
    if ((message.s[i] < 32) || (message.s[i] > 126))
      message.s[i] = '?';

  buffer_puts(buffer_2,"rblsmtpd: ");
  buffer_puts(buffer_2,ip_env);
  buffer_puts(buffer_2," pid ");
  buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
  buffer_puts(buffer_2,": ");
  buffer_put(buffer_2,message.s,message.len);
  buffer_puts(buffer_2,"\n");
  buffer_flush(buffer_2);

  if (!stralloc_cats(&message,"\r\n")) nomem();

  if (!timeout)
    reject();
  else {
    sig_catch(sig_alarm,drop);
    alarm(timeout);
    greet();
    commands(&in,smtpcommands);
  }
  _exit(0);
}

int main(int argc,char **argv,char **envp)
{
  char *x;
  int opt;
  unsigned long greetdelay = 0;

  ip_init();

  x = env_get("RBLSMTPD");
  if (x) {
    if (!*x)
      decision = 1;
    else if (*x == '-') {
      if (!stralloc_copys(&text,x + 1)) nomem();
      decision = 3;
    }
    else {
      if (!stralloc_copys(&text,x)) nomem();
      decision = 2;
    }
  }

  while ((opt = getopt(argc,argv,"bBcCit:r:a:w:W")) != opteof)
    switch(opt) {
      case 'b': flagrblbounce = 1; break;
      case 'B': flagrblbounce = 0; break;
      case 'c': flagfailclosed = 1; break;
      case 'C': flagfailclosed = 0; break;
      case 'i': flagrblinfo = 1; break;
      case 't': scan_ulong(optarg,&timeout); break;
      case 'r': rbl(optarg); break;
      case 'a': antirbl(optarg); break;
      case 'W': if (!decision) { wait(greetdelay); } break;
      case 'w': if (!decision) { scan_ulong(optarg,&greetdelay); wait(greetdelay); } break;
      default: usage();
    }

  argv += optind;
  if (!*argv) usage();
 
  if (flagrblinfo && decision >= 4)
    rblinfo(); 
  else if (decision >= 2)
    rblsmtpd();
  

  pathexec(argv);
  strerr_die4sys(111,FATAL,"unable to run ",*argv,": ");

  return 0;
}
